File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/deps/range.py
from __future__ import annotations

import re

_VERSION_RE = re.compile(r"^(\d+)\.(\d+)\.(\d+)$")


def parse_range(spec) -> dict:
    if spec is None or spec == "":
        return {"op": "any"}
    s = str(spec).strip()
    if s.startswith("^"):
        m = _VERSION_RE.match(s[1:])
        if not m:
            raise ValueError(f"unparseable range: {spec}")
        M, mi, p = (int(x) for x in m.groups())
        return {"op": "caret", "min": [M, mi, p], "max_exclusive": [M + 1, 0, 0]}
    if s.startswith("~"):
        m = _VERSION_RE.match(s[1:])
        if not m:
            raise ValueError(f"unparseable range: {spec}")
        M, mi, p = (int(x) for x in m.groups())
        return {"op": "tilde", "min": [M, mi, p], "max_exclusive": [M, mi + 1, 0]}
    if _VERSION_RE.match(s):
        m = _VERSION_RE.match(s)
        M, mi, p = (int(x) for x in m.groups())
        return {"op": "exact", "min": [M, mi, p], "max_exclusive": [M, mi, p + 1]}
    lo = re.search(r">=\s*(\d+\.\d+\.\d+)", s)
    hi = re.search(r"<\s*(\d+\.\d+\.\d+)", s)
    if lo and hi:
        a = _VERSION_RE.match(lo.group(1)).groups()
        b = _VERSION_RE.match(hi.group(1)).groups()
        return {
            "op": "bounded",
            "min": [int(a[0]), int(a[1]), int(a[2])],
            "max_exclusive": [int(b[0]), int(b[1]), int(b[2])],
        }
    raise ValueError(f"unparseable range: {spec}")


def _cmp(a, b) -> int:
    for x, y in zip(a, b):
        if x != y:
            return x - y
    return 0


def intersects(a: dict, b: dict) -> bool:
    if a["op"] == "any" or b["op"] == "any":
        return True
    lo = a["min"] if _cmp(a["min"], b["min"]) > 0 else b["min"]
    hi = (
        a["max_exclusive"]
        if _cmp(a["max_exclusive"], b["max_exclusive"]) < 0
        else b["max_exclusive"]
    )
    return _cmp(lo, hi) < 0

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/deps/range.mjs
const VERSION_RE = /^(\d+)\.(\d+)\.(\d+)$/;

/**
 * Parse a small semver range subset:
 *   - "^X.Y.Z"      caret  → [X.Y.Z, X+1.0.0)
 *   - "~X.Y.Z"      tilde  → [X.Y.Z, X.Y+1.0)
 *   - "X.Y.Z"       exact  → [X.Y.Z, X.Y.Z+1)
 *   - ">=A <B"      bounded
 *   - null/""       any    → matches anything
 */
export function parseRange(spec) {
  if (spec == null || spec === "") return { op: "any" };
  const s = String(spec).trim();
  if (s.startsWith("^")) {
    const m = s.slice(1).match(VERSION_RE);
    if (!m) throw new Error(`unparseable range: ${spec}`);
    const [, M, mi, p] = m;
    return {
      op: "caret",
      min: [+M, +mi, +p],
      maxExclusive: [+M + 1, 0, 0],
    };
  }
  if (s.startsWith("~")) {
    const m = s.slice(1).match(VERSION_RE);
    if (!m) throw new Error(`unparseable range: ${spec}`);
    const [, M, mi, p] = m;
    return {
      op: "tilde",
      min: [+M, +mi, +p],
      maxExclusive: [+M, +mi + 1, 0],
    };
  }
  if (VERSION_RE.test(s)) {
    const m = s.match(VERSION_RE);
    const [, M, mi, p] = m;
    return {
      op: "exact",
      min: [+M, +mi, +p],
      maxExclusive: [+M, +mi, +p + 1],
    };
  }
  const lo = s.match(/>=\s*(\d+\.\d+\.\d+)/);
  const hi = s.match(/<\s*(\d+\.\d+\.\d+)/);
  if (lo && hi) {
    const a = lo[1].match(VERSION_RE);
    const b = hi[1].match(VERSION_RE);
    return {
      op: "bounded",
      min: [+a[1], +a[2], +a[3]],
      maxExclusive: [+b[1], +b[2], +b[3]],
    };
  }
  throw new Error(`unparseable range: ${spec}`);
}

const cmp = (a, b) => a[0] - b[0] || a[1] - b[1] || a[2] - b[2];

/**
 * Two half-open ranges [min, maxExclusive) intersect iff
 * max(a.min, b.min) < min(a.maxExclusive, b.maxExclusive).
 * "any" intersects with everything.
 */
export function intersects(a, b) {
  if (a.op === "any" || b.op === "any") return true;
  const lo = cmp(a.min, b.min) > 0 ? a.min : b.min;
  const hi = cmp(a.maxExclusive, b.maxExclusive) < 0 ? a.maxExclusive : b.maxExclusive;
  return cmp(lo, hi) < 0;
}
